### DISCUSSION TITLE
How to setup a development environment for ardino based system development while working with this avr-hal

### DISCUSSION THREAD

**Discussion by Vilayat-Ali** [Rahix/avr-hal#456] [Q&A] [answered]
I usually use arduino IDE to write arduino sketchs as it provides easy interface to compile and load binary onto arduino.
What is the recommended development workflow for rust-based approach but with similar easy and direct utilities to compile, test and load binaries onto arduino board.

**Comment by Rahix** [comment, marked as answer]
Hi,

my suggestion is to start a project using the [`avr-hal-template`](https://github.com/Rahix/avr-hal-template).

Then use your Rust IDE or text editor of choice to write your code.  A normal `cargo run` (or the equivalent button in your IDE) should automatically deploy the program to a connected Arduino using our [`ravedude`](https://crates.io/crates/ravedude).

Just give it a try and if you get stuck, feel free to ask for help :)

**Comment by Vilayat-Ali** [comment]
@Rahix I wish to add a `GPS arduino shield` onto a UNO R3 board and then program everything with sweet [Rust](https://www.rust-lang.org). However I dont know how to add library for that? Is there way I can work with existing libs for `.ino` or arduino sketches and then use it in my code using `avr-hal-template`?

**Reply by Rahix** [reply]
No, you need a Rust library for your GPS module.  If none exists, I'm afraid you'll have to build one yourself.

(Also, please open new discussions for new topics)

**Comment by stappersg** [comment]
On Tue, Nov 28, 2023 at 06:05:33AM -0800, Syed Vilayat Ali Rizvi wrote:
> I wish to ....

Begin
